keps/sig-scheduling/20190221-even-pods-spreading.md

```diff
@@ -0,0 +1,350 @@
+
+#### Story 1 - NodeAffinity
+
+As an application developer, I want my application pods to be scheduled onto
+specific topology domains (via NodeAffinity with spec "app in [zone1, zone2,
+zone3]"), but I don't want them to be stacked too much on one topology. (see
+[#68981](https://github.com/kubernetes/kubernetes/issues/68981))
+
+#### Story 2 - PodAffinity
+
+As an application developer, I want my application pods to co-exist with
+particular pods in the same topology domain (via PodAffnity), and I want them to
+be deployed onto separate nodes (or sub domains) as even as possible.
+
+#### Story 3 - PodAntiAffinity
+
+As an application developer, I want my application pods not to co-exist with
+specific pods (via PodAntiAffinity). But in some cases it'd be favorable to
+tolerate "violating" pods in a manageable way. For example, suppose an app
+(replicas=2) is using PodAntiAffinity and deployed onto a 2-nodes cluster, and
+next the app needs to perform a rolling upgrade, then a third replacement pod is
+created, but it failed to be placed due to lack of resource. In this case,
+
+- if CA is enabled, a new machine will be provisioned to hold the new pod
+  (although old replicas will be deleted afterwards) (see
+  [#40358](https://github.com/kubernetes/kubernetes/issues/40358))
+- if CA is not enabled, it's a deadlock since the replacement pod can't be
+  placed. The only workaround at this moment is to update app strategyType from
+  "RollingUpdate" to "Recreate".
+
+Neither of them is an ideal solution. A promising solution is to give user an
+option to trigger "toleration" mode when the cluster is out of resource. Then in
+aforementioned example, a third pod is "tolerated" to be put onto node1 (or
+node2). But keep it in mind, this behavior is only triggered upon resource
+shortage. For a 3-nodes cluster, the third pod will still be placed onto node3
+(if node3 is capable).
+
+### Risks and Mitigations
+
+Along with this feature, additional inevitable cost will be applied for each
+cycle of pod scheduling. So to mitigate potential performance impact, initial
+implementation will _probably_ limit the semantics of "even spreading" on
+`kubernetes.io/hostname`.
+
+We also need to make sure that our implementation will not have any performance
+penalty for pods that do not use this feature.
+
+## Design Details
+
+Basically there are two ways in API design and implementation:
+
+1. implemented as a "sub feature" inside Affinity and hence new API applies to
+   `pod.spec.affinity`. (referred as [Design A](#design-b) in this doc)
+1. implemented as a standalone feature and new API applies to `pod.spec`.
+   (referred as [Design B](#design-b) in this doc)
+
+### Design A
+
+A new structure called `EvenSpreading` is introduced and it's only effective
+when it's not nil.
+
+```go
+type EvenSpreading struct {
+    // MaxSkew describes the degree of imbalance of pods spreading.
+    // Default value is 1 and 0 is not allowed.
+    MaxSkew int32
```

Comment:
is this easily computable by application developers ?

Reply:
I'd say yes :)

For example, if you can tolerate that one topology can have 3 pods, and one topology has 0 pod, then put it as 3 - it's not an absolute number, just the max **difference** which can be tolerated. And adjust the number higher/lower to be more/less tolerant.

Reply:
so its the max difference in the number of pods between any two domains, where domain is a single topologykey or a combination of many , right ?

Reply:
This is the same question/suggestion on a slice of topologyKey, right?

Reply:
No this just me trying to understand MaxSkew.  I think the below comment is not very clear.
```
    // MaxSkew describes the degree of imbalance of pods spreading.
    // Default value is 1 and 0 is not allowed.
```
Would be better to say, its the max difference between the number of pods in any two zones of a given topology type ? Also maxSkew would loose its meaning if we start defining a slice of topologyKeys

Reply:
Comments updated.

I would put the discussion of slice f otopologyKeys on hold, esp. in initial implementation. And if we want to support that, yes, we should have slice of MaxSkew as well, or use slice of pairs of <MaxSkew, TopologyKey>.